osstats/collector.go

```diff
@@ -0,0 +1,175 @@
+package osstats
+
+import (
+	"context"
+	"runtime"
+	"time"
+
+	lttb "github.com/dgryski/go-lttb"
+	"github.com/harness/lite-engine/engine/spec"
+	"github.com/harness/lite-engine/logger"
+	"github.com/shirou/gopsutil/v3/cpu"
+	"github.com/shirou/gopsutil/v3/mem"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	downsampleCount = 10
+)
+
+func New(ctx context.Context, interval time.Duration) *StatsCollector {
+	return &StatsCollector{
+		ctx:      ctx,
+		log:      logger.FromContext(ctx),
+		interval: interval,
+		doneCh:   make(chan struct{}),
+		stats: &spec.OSStats{
+			MemGraph: &spec.Graph{
+				Xmetric: "seconds",
+				Ymetric: "mem_mb",
+			},
+			CpuGraph: &spec.Graph{
+				Xmetric: "seconds",
+				Ymetric: "cpu_milli",
+			},
+		},
+	}
+}
+
+type StatsCollector struct {
+	ctx        context.Context
+	st         time.Time
+	log        *logrus.Entry
+	interval   time.Duration
+	doneCh     chan struct{}
+	stats      *spec.OSStats
+	memPctSum  float64
+	cpuPctSum  float64
+	cpuTotal   int
+	memTotalMB float64
+}
+
+func (s *StatsCollector) Start() {
+	s.st = time.Now()
+	go s.collectStats()
+}
+
+func (s *StatsCollector) Stop() {
+	close(s.doneCh)
+}
+
+func (s *StatsCollector) Stats() *spec.OSStats {
+	return s.stats
+}
+
+// downsample cpu and memory to n points using LTTB
+func (s *StatsCollector) Aggregate() {
+	if len(s.stats.MemGraph.Points) > 0 {
+		s.stats.AvgMemUsagePct = s.memPctSum / float64(len(s.stats.MemGraph.Points))
```

Comment:
(Optional) Average is not a useful stat. We may use percentile instead. e.g. 90 percentile value for memory and cpu.

Reply:
yeah, makes sense. I'll wait for gathering some data and see what we need to add/improve in terms of data collection in subsequent PRs. will keep this in mind.